File: python/Easy/merge_sorted_lists.py
from typing import Optional

puzzle = 'https://leetcode.com/problems/merge-two-sorted-lists/description/'

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1
        
        node_list = ListNode()
        current = node_list

        while list1 and list2:

            if list1.val < list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next

            current = current.next

        # append any remaining
        if list1:
            current.next = list1
        elif list2:
            current.next = list2

        return node_list.next

# Results
# Runtime 43ms beats 95.44% of python3 submissions
# Memory 16.4mb beats 54.35% of python3 submissions

File: javascript/easy/mergeSortedLists.js
const list1 = [1,2,4];
const list2 = [1,3,4];
// [1,1,2,3,4,4]

// plan:
  // 1. create an array to store sortedLists
  // 2. compare the first item in each array since they are sorted
  // 3. Push the smaller item to the array
  // 4. unshift the array that pushed an element
  // 5. keep comparing until both lists are at 0 length

const mergeSortedLists = (list1, list2) => {
  const mergedArray = [];
  while (list1.length !== 0 && list2.length !== 0) {
    if (list1.length === 0 && list2.length === 0) break;

    if (list1.length === 0) {
      mergeSortedLists.push(list2);
      break;
    };
    if (list2.length === 0) {
      mergeSortedLists.push(list1);
      break;
    };

    if (list1[0] < list2[0]) {
      mergedArray.push(list1.shift());
    } else if (list2[0] < list1[0]) {
      mergedArray.push(list2.shift());   
    } else {
      mergedArray.push(list1.shift());
      mergedArray.push(list2.shift());
    }

  };
  return mergedArray;
};

console.log(mergeSortedLists(list1, list2));

// leetcode requires that list1, and list2 are NodeLists, so you can't use javascripts .shift() method the following code must be run on leetcode: 

/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
 
const mergeTwoLists = (list1, list2) => {
  if (!list1) return list2;
  if (!list2) return list1;

  let mergedHead = null;
  let current = null;

  while (list1 && list2) {
    let temp = null;

    if (list1.val < list2.val) {
      temp = list1;
      list1 = list1.next;
    } else {
      temp = list2;
      list2 = list2.next;
    }

    if (!mergedHead) {
      mergedHead = temp;
      current = temp;
    } else {
      current.next = temp;
      current = current.next;
    }
  }

  if (list1) {
    current.next = list1;
  } else {
    current.next = list2;
  }

  return mergedHead;
};